### PR TITLE
server: fix finding pools for volume migration

### LINKED ISSUE #4400: Find suitable storage for migration regression
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
VMware migration, API
~~~

##### CLOUDSTACK VERSION

~~~
4.14
~~~

##### CONFIGURATION

VMware

##### SUMMARY

When executing the api agains attached volume we observe that none of the pools is marked suitable, a little investigation was done suspecting the following change: 
https://github.com/shapeblue/cloudstack/blob/ovfprops-and-vsphere-adv-together/server/src/main/java/com/cloud/server/ManagementServerImpl.java#L156

```
        //This is an override mechanism so we can list the possible local storage pools that a volume in a shared pool might be able to be migrated to
diskProfile.setUseLocalStorage(true);

```
Before searching for the suitable storage pools,  volume is marked as using local storage just to also find any local storage pools to which volume can be migrated along with the shared storage pools. 

But this is causing regression for zone wide storage pools, while zone wide storage pool allocator searches for any zone wide storage pools the entrance check is 

`if (dskCh.useLocalStorage()) {    return null;}
`which causing zone wide storage pool allocator returning nothing and at the end findStoragePoolsForMigration API shows all zone wide storage pools are “not suitable“.

Below is the screenshot of a volume which is not associated with any storage policy but still showing “vvol5-zone“ as “not suitable“.
image-20201011-195301

### LINKED ISSUE #4692: When on Zone level storage, Cluster level pool are not being offered to a volume with find suitable pools api
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
4.15 
~~~

##### SUMMARY

When a volume is sitting on Zone level, it’s not being offered pools to migrate to with find storagepoolsformigration api.

##### STEPS TO REPRODUCE

~~~
1. Get a volume on zone level storage pool
2. Execute find storage pools api
3. Observe cluster level storage pools are not offered
~~~

##### EXPECTED RESULTS

~~~
Cluster level pools a eligible for migration and should be part of the response
~~~

##### ACTUAL RESULTS

~~~
Only zone level pools are there
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

While finding pools for volume migration list following compatible storages:
- all zone-wide storages of the same hypervisor.
- when the volume is attached to a VM, then all storages from the same cluster as that of VM.
- for detached volume, all storages that belong to clusters of the same hypervisor. 

Fixes #4692 
Fixes #4400 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Using cmk and UI.
Tested with a multi-cluster VMware env.
Storage Pools,
```
(localcloud) SBCM5> > list storagepools filter=id,name,scope,clustername,tags,hypervisor
{
  "count": 8,
  "storagepool": [
    {
      "hypervisor": "VMware",
      "id": "53a4a81b-0271-3989-94d1-a00708fd9c32",
      "name": "ps8",
      "scope": "ZONE",
      "tags": "tag1"
    },
    {
      "hypervisor": "VMware",
      "id": "17899b5b-1838-3e8d-b0fe-76cab34093b1",
      "name": "ps7",
      "scope": "ZONE"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "98a0e79c-f879-36e2-b9ce-230f66e88698",
      "name": "ps6",
      "scope": "CLUSTER",
      "tags": "tag1"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "b9aa5fb3-ce0a-316c-a2bb-3dbf905082d4",
      "name": "ps5",
      "scope": "CLUSTER"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c2",
      "id": "79f5ea41-e6dd-3f9d-acca-10267cf7b796",
      "name": "ps4",
      "scope": "CLUSTER",
      "tags": "tag1"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c2",
      "id": "0a74f7c2-63f6-374f-b5df-bc6304d0ff58",
      "name": "ps2",
      "scope": "CLUSTER"
    },
    {
      "clustername": "p1-c1",
      "id": "e777166e-7dbe-340d-977d-a51d5349b5f1",
      "name": "ps2",
      "scope": "CLUSTER",
      "tags": "tag1"
    },
    {
      "clustername": "p1-c1",
      "id": "b2888e3e-b1e2-36d9-9936-861bc6901d3f",
      "name": "ps1",
      "scope": "CLUSTER"
    }
  ]
}
```

Disk offerings used for data disks,
```
(localcloud) SBCM5> > list diskofferings id=bc71e51d-b1fc-4311-87d6-23b366274038 filter=id,name,tags
{
  "count": 1,
  "diskoffering": [
    {
      "id": "bc71e51d-b1fc-4311-87d6-23b366274038",
      "name": "Tagged",
      "tags": "tag1"
    }
  ]
}
(localcloud) SBCM5> > list diskofferings id=310f053a-1c13-47fe-8590-ef3dd1f74e87 filter=id,name,tags
{
  "count": 1,
  "diskoffering": [
    {
      "id": "310f053a-1c13-47fe-8590-ef3dd1f74e87",
      "name": "Small"
    }
  ]
}
```

Disks tested,

1. Detached volume with tagged offering on a zone-wide pool
2. Detached volume with untagged offering on cluster scope pool
3. Volume attached to a VM with tagged offering on a zone-wide pool
4. Volume attached to a VM with untagged offering on a cluster scope pool
5. Root volume of a VM with untagged compute offering on a cluster scope pool
6. Detached volume with tagged offering on a cluster scope pool

```
(localcloud) SBCM5> > list volumes ids=73d242e1-f8fb-4447-8c78-a1e10d106694,7af032e8-a2b3-49bf-b280-45884a0aed90,4f7e80f1-f3b7-4907-9437-0cc02a14d615,65f5248e-dd27-42c8-83e2-bc1e0b781d27,db7e952a-507f-45f8-b113-724f0f4cc9fc,e665c777-7339-4604-9535-2564c23abca8 filter=id,name,virtualmachineid,vmname,diskofferingid,diskofferingname,storageid,storage,hypervisor
{
  "count": 6,
  "volume": [
    {
      "diskofferingid": "bc71e51d-b1fc-4311-87d6-23b366274038",
      "diskofferingname": "Tagged",
      "hypervisor": "VMware",
      "id": "db7e952a-507f-45f8-b113-724f0f4cc9fc",
      "name": "DATA-5-3",
      "storage": "ps8",
      "storageid": "53a4a81b-0271-3989-94d1-a00708fd9c32"
    },
    {
      "diskofferingid": "310f053a-1c13-47fe-8590-ef3dd1f74e87",
      "diskofferingname": "Small",
      "hypervisor": "VMware",
      "id": "65f5248e-dd27-42c8-83e2-bc1e0b781d27",
      "name": "DATA-5-2",
      "storage": "ps5",
      "storageid": "b9aa5fb3-ce0a-316c-a2bb-3dbf905082d4"
    },
    {
      "diskofferingid": "bc71e51d-b1fc-4311-87d6-23b366274038",
      "diskofferingname": "Tagged",
      "hypervisor": "VMware",
      "id": "4f7e80f1-f3b7-4907-9437-0cc02a14d615",
      "name": "DATA-5-1",
      "storage": "ps8",
      "storageid": "53a4a81b-0271-3989-94d1-a00708fd9c32",
      "virtualmachineid": "c893649d-cc11-43f6-81a2-a96a1c0cb9a8",
      "vmname": "t1"
    },
    {
      "diskofferingid": "310f053a-1c13-47fe-8590-ef3dd1f74e87",
      "diskofferingname": "Small",
      "hypervisor": "VMware",
      "id": "7af032e8-a2b3-49bf-b280-45884a0aed90",
      "name": "DATA-5",
      "storage": "ps5",
      "storageid": "b9aa5fb3-ce0a-316c-a2bb-3dbf905082d4",
      "virtualmachineid": "c893649d-cc11-43f6-81a2-a96a1c0cb9a8",
      "vmname": "t1"
    },
    {
      "hypervisor": "VMware",
      "id": "73d242e1-f8fb-4447-8c78-a1e10d106694",
      "name": "ROOT-5",
      "storage": "ps5",
      "storageid": "b9aa5fb3-ce0a-316c-a2bb-3dbf905082d4",
      "virtualmachineid": "c893649d-cc11-43f6-81a2-a96a1c0cb9a8",
      "vmname": "t1"
    },
    {
      "diskofferingid": "bc71e51d-b1fc-4311-87d6-23b366274038",
      "diskofferingname": "Tagged",
      "hypervisor": "VMware",
      "id": "e665c777-7339-4604-9535-2564c23abca8",
      "name": "DATA-5-4",
      "storage": "ps6",
      "storageid": "98a0e79c-f879-36e2-b9ce-230f66e88698"
    }
  ]
}
```

findStoragePoolForMigration results in above order:
```
(localcloud) SBCM5> > find storagepoolsformigration id=db7e952a-507f-45f8-b113-724f0f4cc9fc filter=id,name,scope,clustername,tags,hypervisor,suitableformigration
{
  "count": 7,
  "storagepool": [
    {
      "clustername": "p1-c1",
      "id": "e777166e-7dbe-340d-977d-a51d5349b5f1",
      "name": "ps2",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c2",
      "id": "79f5ea41-e6dd-3f9d-acca-10267cf7b796",
      "name": "ps4",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "98a0e79c-f879-36e2-b9ce-230f66e88698",
      "name": "ps6",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "hypervisor": "VMware",
      "id": "17899b5b-1838-3e8d-b0fe-76cab34093b1",
      "name": "ps7",
      "scope": "ZONE",
      "suitableformigration": false
    },
    {
      "clustername": "p1-c1",
      "id": "b2888e3e-b1e2-36d9-9936-861bc6901d3f",
      "name": "ps1",
      "scope": "CLUSTER",
      "suitableformigration": false
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c2",
      "id": "0a74f7c2-63f6-374f-b5df-bc6304d0ff58",
      "name": "ps2",
      "scope": "CLUSTER",
      "suitableformigration": false
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "b9aa5fb3-ce0a-316c-a2bb-3dbf905082d4",
      "name": "ps5",
      "scope": "CLUSTER",
      "suitableformigration": false
    }
  ]
}

(localcloud) SBCM5> > find storagepoolsformigration id=65f5248e-dd27-42c8-83e2-bc1e0b781d27 filter=id,name,scope,clustername,tags,hypervisor,suitableformigration
{
  "count": 7,
  "storagepool": [
    {
      "clustername": "p1-c1",
      "id": "b2888e3e-b1e2-36d9-9936-861bc6901d3f",
      "name": "ps1",
      "scope": "CLUSTER",
      "suitableformigration": true
    },
    {
      "clustername": "p1-c1",
      "id": "e777166e-7dbe-340d-977d-a51d5349b5f1",
      "name": "ps2",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c2",
      "id": "0a74f7c2-63f6-374f-b5df-bc6304d0ff58",
      "name": "ps2",
      "scope": "CLUSTER",
      "suitableformigration": true
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c2",
      "id": "79f5ea41-e6dd-3f9d-acca-10267cf7b796",
      "name": "ps4",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "98a0e79c-f879-36e2-b9ce-230f66e88698",
      "name": "ps6",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "hypervisor": "VMware",
      "id": "17899b5b-1838-3e8d-b0fe-76cab34093b1",
      "name": "ps7",
      "scope": "ZONE",
      "suitableformigration": true
    },
    {
      "hypervisor": "VMware",
      "id": "53a4a81b-0271-3989-94d1-a00708fd9c32",
      "name": "ps8",
      "scope": "ZONE",
      "suitableformigration": true,
      "tags": "tag1"
    }
  ]
}

(localcloud) SBCM5> > find storagepoolsformigration id=4f7e80f1-f3b7-4907-9437-0cc02a14d615 filter=id,name,scope,clustername,tags,hypervisor,suitableformigration
{
  "count": 3,
  "storagepool": [
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "98a0e79c-f879-36e2-b9ce-230f66e88698",
      "name": "ps6",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "hypervisor": "VMware",
      "id": "17899b5b-1838-3e8d-b0fe-76cab34093b1",
      "name": "ps7",
      "scope": "ZONE",
      "suitableformigration": false
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "b9aa5fb3-ce0a-316c-a2bb-3dbf905082d4",
      "name": "ps5",
      "scope": "CLUSTER",
      "suitableformigration": false
    }
  ]
}

(localcloud) SBCM5> > find storagepoolsformigration id=7af032e8-a2b3-49bf-b280-45884a0aed90 filter=id,name,scope,clustername,tags,hypervisor,suitableformigration
{
  "count": 3,
  "storagepool": [
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "98a0e79c-f879-36e2-b9ce-230f66e88698",
      "name": "ps6",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "hypervisor": "VMware",
      "id": "17899b5b-1838-3e8d-b0fe-76cab34093b1",
      "name": "ps7",
      "scope": "ZONE",
      "suitableformigration": true
    },
    {
      "hypervisor": "VMware",
      "id": "53a4a81b-0271-3989-94d1-a00708fd9c32",
      "name": "ps8",
      "scope": "ZONE",
      "suitableformigration": true,
      "tags": "tag1"
    }
  ]
}

(localcloud) SBCM5> > find storagepoolsformigration id=73d242e1-f8fb-4447-8c78-a1e10d106694 filter=id,name,scope,clustername,tags,hypervisor,suitableformigration
{
  "count": 3,
  "storagepool": [
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "98a0e79c-f879-36e2-b9ce-230f66e88698",
      "name": "ps6",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "hypervisor": "VMware",
      "id": "17899b5b-1838-3e8d-b0fe-76cab34093b1",
      "name": "ps7",
      "scope": "ZONE",
      "suitableformigration": true
    },
    {
      "hypervisor": "VMware",
      "id": "53a4a81b-0271-3989-94d1-a00708fd9c32",
      "name": "ps8",
      "scope": "ZONE",
      "suitableformigration": true,
      "tags": "tag1"
    }
  ]
}

(localcloud) SBCM5> > find storagepoolsformigration id=e665c777-7339-4604-9535-2564c23abca8 filter=id,name,scope,clustername,tags,hypervisor,suitableformigration
{
  "count": 7,
  "storagepool": [
    {
      "clustername": "p1-c1",
      "id": "e777166e-7dbe-340d-977d-a51d5349b5f1",
      "name": "ps2",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c2",
      "id": "79f5ea41-e6dd-3f9d-acca-10267cf7b796",
      "name": "ps4",
      "scope": "CLUSTER",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "hypervisor": "VMware",
      "id": "53a4a81b-0271-3989-94d1-a00708fd9c32",
      "name": "ps8",
      "scope": "ZONE",
      "suitableformigration": true,
      "tags": "tag1"
    },
    {
      "hypervisor": "VMware",
      "id": "17899b5b-1838-3e8d-b0fe-76cab34093b1",
      "name": "ps7",
      "scope": "ZONE",
      "suitableformigration": false
    },
    {
      "clustername": "p1-c1",
      "id": "b2888e3e-b1e2-36d9-9936-861bc6901d3f",
      "name": "ps1",
      "scope": "CLUSTER",
      "suitableformigration": false
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c2",
      "id": "0a74f7c2-63f6-374f-b5df-bc6304d0ff58",
      "name": "ps2",
      "scope": "CLUSTER",
      "suitableformigration": false
    },
    {
      "clustername": "10.10.3.184/Trillian/p1-c3",
      "id": "b9aa5fb3-ce0a-316c-a2bb-3dbf905082d4",
      "name": "ps5",
      "scope": "CLUSTER",
      "suitableformigration": false
    }
  ]
}
```

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/StoragePoolAllocator.java [hunk 1]
```diff
@@ -45,10 +51,12 @@
      * @param ExcludeList
      *            avoid
      * @param int returnUpTo (use -1 to return all possible pools)
+     * @param boolean bypassStorageTypeCheck
```
Comment (DaanHoogland) [inline engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/StoragePoolAllocator.java:54]: why add javadoc without description? please add a short sentence on how to use it, or remove it.

**Review by DaanHoogland — CHANGES_REQUESTED** [review]
clgtm, javadox should make sense

File: engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/StoragePoolAllocator.java (hunk 1 above)
Comment (shwstppr) [reply engine/api/src/main/java/org/apache/cloudstack/engine/subsystem/api/storage/StoragePoolAllocator.java]: @DaanHoogland done

**Review by DaanHoogland — APPROVED** [review]

**Review by rhtyd — APPROVED** [review]

**Pull request merged**